Gd 190

Barmalej24 opened the pull request:
Уровень домена
1. Реализован абстрактный класс StringValueObject
2. StringValueObject унаследован всем ValueObject
3. Добавлена новая сущность для работы с именами очередей (соотношение очереди с типом шаблона уведомления)
4. Во всех сущностях максимально используются реализованные ValueObject
5. Добавлен enum для языка шаблона
6. Удален enum для событий

Уровень Data
1. Реализован базовый абстрактный репозиторий
2. Реализован интерфейс репозитория для новой доменной сущности очередей
3. Реализован репозиторий для новой доменной сущности очередей
4. Внесены изменения (добавлено наследование базового репозитория) во все ранее существовавшие репозитории сущностей
5. Добавлена конфигурация EF для новой доменной сущности

Уровень приложения
1. Добавлен маркерный интерфейс для моделей создания
2. Во все интерфейсы сервисов и реализации сервисов добавлены токены отмены
3. Добавлен интерфейс для сервиса работы с новой доменной сущностью
4. Реализован сервис по работе с новой доменной сущностью
5. Добавлены новые модели для работы с новой доменной сущностью

Уровень представления
1. Добавлен сервис для отправки сообщений в очереди (в сторону Email сервиса)
2. Добавлен сервис принятия сообщений из очереди от сервисов Регистрации и Профиля пользователя
3. Добавлен котроллер для работы с новой доменной сущностью 
4. Добавлены контракты для новой доменной сущности
5. Изменен валидатор для языка шаблона

Barmalej24 committed:
Added  Enum Direction, Event
Added valueobj FullName
Added IAddRepository
Added new services RMQProducer and Send
Changed TemplateService (added metod GetByTypeId)
Changed MessageController for try send EmailConfirmation event
Fixed Exception

Barmalej24 committed:
Allocation of a separate interface for events
Removal of a separate event service

Barmalej24 committed:
added fluentvalidation for events

Barmalej24 committed:
Domain Level
* Implemented abstract StringValueObject class
* Inherited StringValueObject by all ValueObjects
* Added new entity for working with queue names (queue-template type relationship)
* Maximized usage of implemented ValueObjects in all entities
* Added enum for template language
* Removed enum for events

Data Level
* Implemented base abstract repository
* Implemented interface for new domain entity repository (queue)
* Implemented repository for new domain entity (queue)
* Made changes (added inheritance from base repository) to all existing entity repositories
* Added EF configuration for new domain entity

Application Level
* Added marker interface for creation models
* Added cancellation tokens to all service interfaces and implementations
* Added interface for service working with new domain entity
* Implemented service for working with new domain entity
* Added new models for working with new domain entity

Presentation Level
* Added service for sending messages to queues (towards Email service)
* Added service for receiving messages from queue from Registration and Profile services
* Added controller for working with new domain entity
* Added contracts for new domain entity
* Changed validator for template language

Barmalej24 committed:
Chenged QueueName InValid
Added QueueName Exception

Barmalej24 committed:
Added BusQueue Validator
Fix QueueName Exception

Barmalej24 committed:
Added logic get default template

Barmalej24 committed:
Add Docker

oleggolen commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+      POSTGRES_PASSWORD: "admin"
+    volumes:
+      - local_pgdata:/var/lib/postgresql/data
+
```
необходимо добавить health checks для базы данных

Barmalej24 replied:
добавил

oleggolen commented on docker-compose.yml:
```diff
@@ -0,0 +1,24 @@
+    ports:
+      - "3000:8080"
+    depends_on:
+      - db
```
данные подключения к базе данных лучше передавать в контейнер бэкэнда переменными окружения

oleggolen commented on NotificationMicroservice.Domain/Entities/MessageTemplate.cs:
```diff
@@ -22,7 +23,7 @@ public class MessageTemplate : IModifyEntity<User, Guid>
         /// <summary>
         /// Максимальная длина названия для шаблона сообщения
         /// </summary>
-        public const int TEMPLATE_MAX_LENGTH = 250;
+        public const int TEMPLATE_MAX_LENGTH = 600;
```
Я бы не ограничивал длину шаблона сообщения

Barmalej24 replied:
убрал

oleggolen commented on NotificationMicroservice.DataAccess/Configuration/TemplateConfiguration.cs:
```diff
@@ -15,9 +17,16 @@ public void Configure(EntityTypeBuilder<MessageTemplate> builder)
+                    s => (Language)Enum.Parse(typeof(Language), s))
                 .IsRequired();
 
             builder.Property(x => x.Template)
+                .HasMaxLength(MessageTemplate.TEMPLATE_MAX_LENGTH)
+                .HasConversion(
+                    o => o.Value,
+                    s => new Template(s))
                 .IsRequired();
```
Я бы не ограничевал длину шаблона сообщения

Barmalej24 replied:
убрал.

oleggolen commented on NotificationMicroservice.DataAccess/Configuration/MessageConfiguration.cs:
```diff
@@ -14,6 +15,9 @@ public void Configure(EntityTypeBuilder<Message> builder)
                 .ValueGeneratedOnAdd();
 
             builder.Property(x => x.Text)
+                .HasConversion(
+                    o => o.Value,
+                    s => new MessageText(s))
```
Я бы не ограничевал длину сообщения

Barmalej24 replied:
сообщение не ограничено по длине

oleggolen commented on NotificationMicroservice.Application/Services/MessageTemplateService.cs:
```diff
@@ -3,88 +3,101 @@
+
+            if (dbEntities is null)
+            {
+                return null;
```
Я бы здесь не возвращал null, а вернул бы пустую коллекцию, да и условие никогда не выполнится, так как из оепозитория возвращается тоже пустая коллекция, а не nullable тип

Barmalej24 replied:
убрал

oleggolen commented on NotificationMicroservice.Application/Model/User/UserModel.cs:
```diff
@@ -1,4 +1,6 @@
+
+namespace NotificationMicroservice.Application.Model.User
 {
     public class UserModel : IBaseModel<Guid>
     {
```
Лучше просто все поля сделать required или хотя бы некоторые, такие как UserName и Email они гарантированно должны приходить

Barmalej24 replied:
преобразовал в запись

oleggolen commented on NotificationMicroservice.Application/Model/Type/UpdateTypeModel.cs:
```diff
@@ -1,4 +1,6 @@
+
+namespace NotificationMicroservice.Application.Model.Type
 {
     public class UpdateTypeModel : IBaseModel<Guid>
     {
```
Лучше сделать всё required

Barmalej24 replied:
добавил

oleggolen commented on NotificationMicroservice.Application/Model/Type/TypeModel.cs:
Лучше сделать все свойства required

Barmalej24 replied:
преобразовал в запись

oleggolen commented on NotificationMicroservice.Application/Model/Type/DeleteTypeModel.cs:
Все свойства сделать required

Barmalej24 replied:
добавил

oleggolen commented on NotificationMicroservice.Application/Model/Template/UpdateTemplateModel.cs:
Лучше все поля сделать required

Barmalej24 replied:
добавил

oleggolen commented on NotificationMicroservice.Application/Model/Template/TemplateModel.cs:
Лучше все поля сделать required кроме nullable

Barmalej24 replied:
преобразовал в запись

oleggolen commented on NotificationMicroservice.Application/Model/Template/DeleteTemplateModel.cs:
Лучше все поля сделать required

Barmalej24 replied:
добавил

oleggolen commented on NotificationMicroservice.Application/Model/BusQueue/UpdateBusQueueModel.cs:
Лучше все поля сделать required

Barmalej24 replied:
добавил

oleggolen commented on NotificationMicroservice.Application/Model/BusQueue/DeleteBusQueueModel.cs:
Лучше все поля сделать required

Barmalej24 replied:
добавил

oleggolen commented on NotificationMicroservice.Application/Model/Abstractions/IBaseModel.cs:
```diff
@@ -1,4 +1,4 @@
-﻿namespace NotificationMicroservice.Application.Model
+﻿namespace NotificationMicroservice.Application.Model.Abstractions
 {
     public interface IBaseModel<Tkey> where Tkey : struct
```
Добавить ещё, что ключи сравниваемые

Barmalej24 replied:
добавил

oleggolen commented on NotificationMicroservice/Validator/Queue/EditBusQueueValidator.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class EditBusQueueValidator : AbstractValidator<EditBusQueueRequest>
+    {
+        public EditBusQueueValidator()
+        {
+            RuleFor(type => type.MessageTypeId)
+                    .SetValidator(new GuidValidator());
+
+            RuleFor(type => type.ModifiedByUserId)
+                    .SetValidator(new GuidValidator());
+        }
```
Лишний класс

oleggolen commented on NotificationMicroservice/Validator/Queue/DeleteBusQueueValidator.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace NotificationMicroservice.Validator.Queue
+{
+    public class DeleteBusQueueValidator : AbstractValidator<DeleteBusQueueRequest>
+    {
+        public DeleteBusQueueValidator()
+        {
+            RuleFor(type => type.ModifiedByUserId)
+                    .SetValidator(new GuidValidator());
+        }
+    }
```
Лишний класс

oleggolen commented on NotificationMicroservice/Validator/Queue/AddBusQueueValidator.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public class AddBusQueueValidator : AbstractValidator<AddBusQueueRequest>
+    {
+        public AddBusQueueValidator()
+        {
+
+            RuleFor(type => type.QueueName)
+                .SetValidator(new QueueNameValidator());
+
+            RuleFor(type => type.MessageTypeId)
+                .SetValidator(new GuidValidator());
+
+            RuleFor(type => type.CreatedByUserId)
+                    .SetValidator(new GuidValidator());
+        }
```
Лишний класс

oleggolen commented on NotificationMicroservice/Controllers/BusQueueController.cs:
```diff
@@ -0,0 +1,80 @@
+            return queue is null
+                ? NotFound($"Type not found for QueueName:{queueName}")
+                : Ok(mapper.Map<BusQueueResponse>(queue));
+        }
+
+        [HttpPost]
+        [ProducesResponseType(typeof(Guid), 201)]
+        [ProducesResponseType(typeof(string), 400)]
+        public async Task<ActionResult<Guid>> AddAsync([FromBody] AddBusQueueRequest request, CancellationToken cancellationToken)
+        {
+            var queueId = await queueService.AddAsync(mapper.Map<CreateBusQueueModel>(request), cancellationToken);
+
+            if (queueId is null)
+            {
+                return BadRequest("BusQueue can not be created");
+            }
+            return Created("", queueId);
+        }
+
+        [HttpPut("{id:guid}")]
+        [ProducesResponseType(204)]
+        [ProducesResponseType(typeof(bool), 400)]
+        public async Task<ActionResult<bool>> UpdateAsync(Guid id, [FromBody] EditBusQueueRequest request, CancellationToken cancellationToken)
+        {
+            var queue = mapper.Map<UpdateBusQueueModel>(request);
+            queue.Id = id;
+            return await queueService.UpdateAsync(queue, cancellationToken) is true ? NoContent() : BadRequest(false);
+        }
+
+        [HttpDelete("{id:guid}")]
+        [ProducesResponseType(204)]
+        [ProducesResponseType(typeof(bool), 400)]
+        public async Task<ActionResult<bool>> DeleteAsync(Guid id, [FromBody] DeleteBusQueueRequest request, CancellationToken cancellationToken)
+        {
+            var queue = mapper.Map<DeleteBusQueueModel>(request);
+            queue.Id = id;
+            return await queueService.DeleteAsync(queue, cancellationToken) is true ? NoContent() : BadRequest(false);
+        }
+    }
+}
```
Лишний контроллер

oleggolen commented on NotificationMicroservice/Validator/Base/QueueNameValidator.cs:
```diff
@@ -0,0 +1,16 @@
+                .NotEmpty()
+                .Length(BusQueue.MIN_QUEUENAME_LENGTH, BusQueue.MAX_QUEUENAME_LENGTH);
+        }
+    }
```
Лишний класс

oleggolen commented on NotificationMicroservice/Validator/Event/UpdateUserValidator.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    public class UpdateUserValidator : AbstractValidator<UpdateUserEvent>
+    {
+        public UpdateUserValidator()
+        {
+            RuleFor(user => user.Id)
+                .SetValidator(new GuidValidator());
+            RuleFor(user => user.Email)
+                .SetValidator(new EmailValidator());
+            RuleFor(user => user.FullName)
+                .SetValidator(new FullNameValidator());
+        }
```
Лишний класс

Barmalej24 replied:
удалил.

oleggolen commented on NotificationMicroservice/Validator/Event/CreateUserValidator.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public class CreateUserValidator : AbstractValidator<CreateUserEvent>
+    {
+        public CreateUserValidator()
+        {
+            RuleFor(user => user.Id)
+                .SetValidator(new GuidValidator());
+            RuleFor(user => user.Email)
+                .SetValidator(new EmailValidator());
+            RuleFor(user => user.Username)
+                .SetValidator(new FullNameValidator());
+            RuleFor(user => user.FullName)
+                .SetValidator(new FullNameValidator());
+        }
```
Лишний класс

Barmalej24 replied:
удалил.

oleggolen commented on NotificationMicroservice/Validator/Event/ConfirmationEmailValidator.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace NotificationMicroservice.Validator.Event
+{
+    public class ConfirmationEmailValidator : AbstractValidator<ConfirmationEmailEvent>
+    {
+        public ConfirmationEmailValidator()
+        {
+            RuleFor(request => request.Email)
+                .SetValidator(new EmailValidator());
+            RuleFor(request => request.Username)
+                .SetValidator(new FullNameValidator());
+            RuleFor(request => request.Link)
+                .SetValidator(new UriValidator());
+        }
+    }
```
Лишний класс

Barmalej24 replied:
удалил.

oleggolen commented on NotificationMicroservice/Services/NotificationControlService.cs:
```diff
@@ -0,0 +1,80 @@
+                    var updateUser = JsonSerializer.Deserialize<UpdateUserEvent>(message);
+                    return await userService.UpdateAsync(mapper.Map<UpdateUserModel>(updateUser), cancellationToken);
+            }
+            return false;
```
Как будто бы здесь лучше всего использовать паттерн команда и фабричный метод для этих команд, потому что так выглядит очень сильно, либо использовать фабричный метод и MediatR

oleggolen commented on NotificationMicroservice/Services/NotificationControlService.cs:
```diff
@@ -0,0 +1,80 @@
+            return false;
+        }
+
+        private async Task<TemplateModel?> GetTemplate(string queue, CancellationToken cancellationToken)
+        {
+            var busQueue = await queueApplicationService.GetTypeByEventAsync(queue, cancellationToken);
+
+            if (busQueue == null)
+            {
+                return null;
+            }
+
+            var templates = await templateService.GetByTypeIdAsync(busQueue.Type.Id, cancellationToken);
+
+            if (templates == null)
+            {
+                return null;
+            }
+
+            var template = templates.FirstOrDefault(i => i.Language == _lang);
+
+            if (template == null)
+            {
+                template = templates.FirstOrDefault(i => i.Language == "Eng");
+            }
+
+            return template is null ? null : template;
+        }
```
Вот тут можно попробовать в репозиториях идею с IQuerible или дописать репозиторию конкретный метод с выборкой по языку и типу сообщения, чтобы не писать вот такие килмометровые портянки выборки из БД на уровне приложения

Barmalej24 replied:
вынес в репозиторий

oleggolen commented on NotificationMicroservice/Contracts/BusQueue/DeleteBusQueueRequest.cs:
```diff
@@ -0,0 +1,4 @@
+﻿namespace NotificationMicroservice.Contracts.BusQueue
+{
+    public record DeleteBusQueueRequest(Guid ModifiedByUserId);
```
Лишний класс

oleggolen reviewed (changes requested):
Тут у меня промелькнула идея переписать это всё дело вообще на чистую архитектуру, потому что мне очень не нравятся классы, в которых упомянаются ифраструктурные вещи типа очереди сообщений в домене, ну вот вообще они не имеют никакого отношения к предметной области решаемой задачи, а так будут эти классы только в инфраструктуре, и репозитории для работы с ними будут там же и никакой домент и бизнес-логика не будут знать ничего про то, что там по брокерам гоняется, как собственно и должно быть. Также ещё несколько моментов по аккуратности кода и контейнеризации., ну и поубирать лишние контроллеры,  и для низ DTO лучше это всё сразу через брокер тестировать. Кстати брокер я бы тоже внёс бы в docker-compose файл, на него бы навесил переменные и health checks. В общем надо подумать, как лучше сделать, но у меня соображения такие

Barmalej24 committed:
Domain :: Removed parameter for maximum long text of message template, made changes to entity, valueobj, exceptions
Application :: Base classes for models have been converted into records, properties of the action model field have been changed
Repository :: Added two methods to search for templates by message type and language tag

Barmalej24 committed:
Deleted BusQueue from Domain
